takashi-dependency-manager.md

```diff
@@ -0,0 +1,108 @@
+also allow for users to install dependencies from a particular git
+branch, or particular upstream (maybe they don't trust our centralized
+one, and want to use their own private registry, like self-hosted npm or
+maven).
+
+## Questions
+
+1. How do we guarantee to our users that their downloads haven't been
+   tampered with before they make it to their computer?
+2. How do we deal with users using different architectures and operating
+   systems? (brew runs on x86_64, AArch64, and more architectures), and
+   supports mac and linux for operating systems.
+3. Should we allow direct binary downloads, or should we force all of
+   our users to compile from source? Why is this important?
+4. How do dependency packagers express that their package depends on
+   certain things? How is this fraught with peril? What are some things
+   to keep in mind while doing this?
+
+## Features
+
+Let's consider the three sides of the project:
+
+1. Client
+2. Maintainers
+3. Backend (CDN)
+
+### Client
+
+The client should be able to do a few tasks:
+
+1. Downloading a dependency
+2. Query for downloaded dependencies
+3. Remove a dependency
+
+#### Downloading a dependency
+
+To download a dependency, the client will connect to a CDN and download
+either the source code (and compile the dependency on their machine) or
+download a binary (if compiled) or the source code (if interpreted)
+after checking for dependencies.
+
+To download the required dependencies, the client should check the
+dependencies required, and topological sort to resolve a dependency
+chain required for the dependency.
+
+It's important to make sure that the dependency works on the
+architecture + OS + libc (target triplet) and all dependencies do as
+well, before trying to download the target dependency.
+
+The question about downloading source code or a binary is an important
+design decision:
+
+1. If we allow our users to download arbitrary source code and then
+   compile code, package maintainers must audit the source code to make
+   sure that it is correct.
+2. If we allow package maintainers to post binaries, we need to audit
+   binaries to make sure they're secure.
+
+(We could allow the best of both worlds by having maintainers audit the
+code of dependencies and then compile that and post that to the CDN).
+
+However, we have some problems either way: since there are some critical
+packages (those which are depended upon by lots of dependencies), a hack
+into one of those will trickle down to affect all of our users, so
+security is extremely important.
+
+### Maintainers
+
+Maintainers should be able to express which dependencies that their
+dependencies require (e.g. dynamically linked libraries to run a
+binary). They should be able to express a range of versions that these
+dependencies can run on, and test this in CI to make sure that this
+doesn't break some users. Tests should be required for this.
+
+### Backend
```

Comment:
If there is something wrong with someones code that is causing a security vulnerablity how can get the maintainer to address it? What sorts of aggressive tactics can we take ?